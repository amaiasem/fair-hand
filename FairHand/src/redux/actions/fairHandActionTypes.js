export default {
  LOAD_ALL_SHOPS: 'LOAD_ALL_SHOPS'
}
