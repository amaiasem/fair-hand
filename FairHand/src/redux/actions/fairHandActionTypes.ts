export default {
  LOAD_ALL_SHOPS: 'LOAD_ALL_SHOPS',
  FILTER_SHOP_BY_TYPE: 'FILTER_SHOP_BY_TYPE',
  FILTER_SHOP_BY_NAME: 'FILTER_SHOP_BY_NAME',
  GET_REVIEWS_BY_SHOP_NAME: 'GET_REVIEWS_BT_SHOP_NAME',
  GET_REVIEWS_BY_USER_NAME: 'GET_REVIEWS_BY_USER_NAME',
  USER_SIGN_IN: 'USER_SIGN_IN',
  USER_REGISTER: 'USER_REGISTER',
  USER_SIGN_OUT: 'USER_SIGN_OUT',
  ADD_REVIEW: 'ADD_REVIEW'
}
