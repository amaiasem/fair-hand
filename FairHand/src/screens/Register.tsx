import React, { useState } from 'react'
import { StyleSheet, Text, View, Image, TouchableOpacity } from 'react-native'
import { Ionicons } from '@expo/vector-icons'
import { COLOR, SIZES, images } from '../../constants'
import { TextInput } from 'react-native-gesture-handler'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 30
  },
  goBack: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: SIZES.width,
    padding: 30
  },
  icon: {
    fontSize: 50,
    color: COLOR.lightgrey
  },
  loginForm: {
    flex: 0.5,
    width: SIZES.width,
    padding: 30,
    justifyContent: 'center',
    alignItems: 'flex-start'
  },
  title: {
    fontSize: SIZES.h1,
    fontWeight: '700',
    marginBottom: 20,
    marginTop: 80
  },
  input: {
    width: SIZES.width * 0.85,
    height: 40,
    marginTop: 15,
    paddingLeft: 30,
    borderRadius: SIZES.buttonRadius,
    borderColor: COLOR.grey,
    borderWidth: 1
  },
  button: {
    width: SIZES.width * 0.85,
    height: 40,
    marginTop: 20,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: COLOR.black,
    borderRadius: SIZES.buttonRadius
  },
  buttonText: {
    color: COLOR.white,
    fontSize: SIZES.buttonText,
    fontWeight: '700'
  }
})

const Register = ({ navigation }:any) => {
  const [value, onChangeText] = useState('')
  return (
    <View style={styles.container}>
      <TouchableOpacity style={styles.goBack}>
        <Ionicons
          style={styles.icon}
          name="chevron-back-circle-outline"
          size={24}
          color="black"
          onPress={() => navigation.goBack()}
          />
      </TouchableOpacity>
      <Image
        source= {images.orangeLogo}
      ></Image>
      <View style={styles.loginForm}>
        <Text style={styles.title}>Create your account</Text>
        <TextInput
          style={styles.input}
          onChangeText={text => onChangeText(text)}
          placeholder='Name'
          value = {value}
        ></TextInput>
        <TextInput
          style={styles.input}
          onChangeText={text => onChangeText(text)}
          placeholder='Surname'
          value = {value}
        ></TextInput>
        <TextInput
          style={styles.input}
          onChangeText={text => onChangeText(text)}
          placeholder='Email'
          value = {value}
        ></TextInput>
        <TextInput
          style={styles.input}
          onChangeText={text => onChangeText(text)}
          placeholder='Password'
          value = {value}
        ></TextInput>
        <TouchableOpacity style={styles.button}>
            <Text style={styles.buttonText}>SUBMIT</Text>
        </TouchableOpacity>
      </View>
    </View>
  )
}

export default Register
