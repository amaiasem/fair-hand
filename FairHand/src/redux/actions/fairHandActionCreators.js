import axios from 'axios'
import fairHandActionTypes from './fairHandActionTypes'

export default function loadAllShops () {
  return async function fetchInfo (dispatch) {
    const { data } = await axios('http://192.168.0.41:5000/shops')
    dispatch({
      type: fairHandActionTypes.LOAD_ALL_SHOPS,
      data
    })
  }
}

export function filterShopsByType (shops, typeClothes) {
  const filteredShops = shops.filter(shop => shop.type.includes(typeClothes) === true)

  return {
    type: fairHandActionTypes.FILTER_SHOP_BY_TYPE,
    data: filteredShops
  }
}

export function filterShopsByName (shops, shopName) {
  const filteredShops = shops.filter(shop => shop.shopName.includes(shopName) === true)

  return {
    type: fairHandActionTypes.FILTER_SHOP_BY_NAME,
    data: filteredShops
  }
}
