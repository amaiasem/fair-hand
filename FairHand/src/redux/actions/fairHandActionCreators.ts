import axios from 'axios'
import fairHandActionTypes from './fairHandActionTypes'
import ShopInterface from '../../Interfaces/shopInterface'
import Review from '../../Interfaces/reviewInterface'

// Shop actions

export function loadAllShops () {
  return async function fetchInfo (dispatch: any) {
    const { data } = await axios.get('http://192.168.0.41:5000/shops')
    dispatch({
      type: fairHandActionTypes.LOAD_ALL_SHOPS,
      data
    })
  }
}

export function filterShopsByType (shops: ShopInterface[], typeClothes: string) {
  const filteredShops = shops.filter(shop => shop.type.includes(typeClothes) === true)

  return {
    type: fairHandActionTypes.FILTER_SHOP_BY_TYPE,
    data: filteredShops
  }
}

export function filterShopsByName (shops: ShopInterface[], shopName: string) {
  if (shopName === '') {
    return {
      type: fairHandActionTypes.FILTER_SHOP_BY_NAME,
      data: shops
    }
  } else {
    const filteredShops = shops.filter(shop => shop.shopName.includes(shopName) === true)

    return {
      type: fairHandActionTypes.FILTER_SHOP_BY_NAME,
      data: filteredShops
    }
  }
}

// Reviews actions

export function getReviewsByShopName (searchShop:string) {
  return async function fetchInfo (dispatch: any) {
    const { data } = await axios.get(`http://192.168.0.41:5000/reviews/shopName/${searchShop}`)

    dispatch({
      type: fairHandActionTypes.GET_REVIEWS_BY_SHOP_NAME,
      data
    })
  }
}

export function getReviewsByUserName (searchName:string) {
  return async function fetchInfo (dispatch: any) {
    const { data } = await axios.get(`http://192.168.0.41:5000/reviews/userName/${searchName}`)

    dispatch({
      type: fairHandActionTypes.GET_REVIEWS_BY_USER_NAME,
      data
    })
  }
}

// User actions

export function userLogin (user: object) {
  return async (dispatch: any) => {
    try {
      const { data } = await axios.post('http://192.168.0.41:5000/auth/login', user)
      dispatch({
        type: fairHandActionTypes.USER_SIGN_IN,
        data
      })
    } catch (error) {
      dispatch({
        type: fairHandActionTypes.USER_SIGN_IN,
        data: 400
      })
    }
  }
}

export function userRegister (user: object) {
  return async (dispatch: any) => {
    const { data } = await axios.post('http://192.168.0.41:5000/auth/register', user)
    if (data === 'User already exists!') {
      dispatch({
        type: fairHandActionTypes.USER_REGISTER
      })
    } else {
      dispatch({
        type: fairHandActionTypes.USER_REGISTER,
        data
      })
    }
  }
}

export function addReview (review: Review) {
  return async (dispatch: any) => {
    try {
      const { data } = await axios.post('http://192.168.0.41:5000/reviews', review)
      dispatch({
        type: fairHandActionTypes.ADD_REVIEW,
        data
      })
    } catch (error) {
      dispatch({
        type: fairHandActionTypes.ADD_REVIEW,
        data: 400
      })
    }
  }
}

export function deleteReview (allReviews: Review[], reviewID: Review) {
  return async (dispatch: any) => {
    try {
      const { data } = await axios.delete('http://192.168.0.41:5000/reviews', {
        data: {
          _id: reviewID
        }
      })
      dispatch({
        type: fairHandActionTypes.DELETE_REVIEW,
        data: { allReviews: allReviews, deleted: data }
      })
    } catch (error) {
      dispatch({
        type: fairHandActionTypes.DELETE_REVIEW,
        data: 400
      })
    }
  }
}

export function userUpdate (user: object) {
  return async (dispatch: any) => {
    const { data } = await axios.put('http://192.168.0.41:5000/user', user)
    if (data === 'Could not update the user') {
      dispatch({
        type: fairHandActionTypes.USER_UPDATE
      })
    } else {
      dispatch({
        type: fairHandActionTypes.USER_UPDATE,
        data
      })
    }
  }
}
