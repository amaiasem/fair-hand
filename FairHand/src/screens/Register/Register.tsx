import React, { useEffect, useState } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { StyleSheet, Text, View, Image, TouchableOpacity } from 'react-native'
import { Ionicons } from '@expo/vector-icons'
import { COLOR, SIZES, images } from '../../../constants'
import { TextInput } from 'react-native-gesture-handler'
import { userRegister } from '../../redux/actions/fairHandActionCreators'
import User from '../../Interfaces/userInterface'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 30
  },
  goBack: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: SIZES.width,
    padding: 30
  },
  icon: {
    fontSize: 50,
    color: COLOR.lightgrey
  },
  loginForm: {
    flex: 0.5,
    width: SIZES.width,
    padding: 30,
    justifyContent: 'center',
    alignItems: 'flex-start'
  },
  title: {
    fontSize: SIZES.h1,
    fontWeight: '700',
    marginBottom: 20,
    marginTop: 80
  },
  input: {
    width: SIZES.width * 0.85,
    height: 40,
    marginTop: 15,
    paddingLeft: 30,
    borderRadius: SIZES.buttonRadius,
    borderColor: COLOR.grey,
    borderWidth: 1
  },
  button: {
    width: SIZES.width * 0.85,
    height: 40,
    marginTop: 20,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: COLOR.black,
    borderRadius: SIZES.buttonRadius
  },
  buttonText: {
    color: COLOR.white,
    fontSize: SIZES.buttonText,
    fontWeight: '700'
  }
})

const Register = ({ user, action, navigation }: {user: User, action: any, navigation: any}) => {
  const [name, setName] = useState('')
  const [email, setEmail] = useState('')
  const [password, setPassword] = useState('')
  const [repeatPassword, setRepeatPassword] = useState('')
  const emailRegex: any = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
  const passwordRegex: any = /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$/

  function registerAndValidate (name: string, email:string, password: string) {
    if (emailRegex.test(email) === false) {
      alert('Please add a valid email')
    }
    if (passwordRegex.test(password) === false) {
      alert('Please add a valid password! Minimum of 8 characters and at least one letter')
    } else if (password === repeatPassword) {
      if (emailRegex.test(email) === true) {
        alert('Thank you for joining Fairhand!')
        action.userRegister({ name, email, password })
      }
    }
  }

  useEffect(() => {
    if (user.email) {
      navigation.navigate('TabNavigator')
    }
  }, [user.email])

  return (
    <View style={styles.container}>
      <TouchableOpacity style={styles.goBack}>
        <Ionicons
          style={styles.icon}
          name="chevron-back-circle-outline"
          size={24}
          color="black"
          onPress={() => navigation.goBack()}
          testID='go-back'
          />
      </TouchableOpacity>
      <Image
        source= {images.orangeLogo}
      ></Image>
      <View style={styles.loginForm}>
        <Text style={styles.title}>Create your account</Text>
        <TextInput
          style={styles.input}
          onChangeText={(event) => setName(event)}
          placeholder='Name'
          value = {name}
          testID='input-name'
          ></TextInput>
        <TextInput
          style={styles.input}
          onChangeText={(event) => setEmail(event)}
          placeholder='Email'
          value = {email}
          testID='input-email'
          autoCapitalize="none"
          ></TextInput>
        <TextInput
          style={styles.input}
          onChangeText={(event) => setPassword(event)}
          placeholder='Password'
          value = {password}
          testID='input-password'
          autoCapitalize="none"
          secureTextEntry={true}
        ></TextInput>
        <TextInput
          style={styles.input}
          onChangeText={(event) => setRepeatPassword(event)}
          placeholder='Repeat password'
          value = {repeatPassword}
          autoCapitalize="none"
          secureTextEntry={true}
          testID='input-password'
        ></TextInput>
        <TouchableOpacity
        style={styles.button}
        disabled={!name || !email || !password}
        onPress={() => registerAndValidate(name, email, password)}>
            <Text style={styles.buttonText}>SUBMIT</Text>
        </TouchableOpacity>
      </View>
    </View>
  )
}

function mapStateToProps (state: any) {
  return {
    user: state.userReducer.user
  }
}

function mapDispatchToProps (dispatch: any) {
  return {
    action: bindActionCreators({ userRegister }, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Register)
