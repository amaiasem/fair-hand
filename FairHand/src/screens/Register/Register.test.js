import React from 'react'
import { Provider } from 'react-redux'
import { cleanup, render, fireEvent } from '@testing-library/react-native'
import renderer from 'react-test-renderer'
import Register from './Register'
import configureStore from 'redux-mock-store'
import * as action from '../../redux/actions/fairHandActionCreators'
jest.mock('../../redux/actions/fairHandActionCreators')

const mockStore = configureStore()
jest.spyOn(action, 'userRegister').mockReturnValue({ type: '' })

describe('Given a component Login', () => {
  let store
  let component
  let goBack
  let navigate

  beforeEach(() => {
    store = mockStore({
      userReducer: { user: [] }
    })
    goBack = jest.fn()
    navigate = jest.fn()
    component = (<Provider store={store}><Register navigation={{ goBack, navigate }}/></Provider>)
  })

  afterEach(() => cleanup())

  it('It should match the snapshot', () => {
    render(component)
    expect(component).toMatchSnapshot()
  })

  describe('When pressing a goBack button', () => {
    it('It should call navigation.goBack()', () => {
      const { getByTestId } = render(component)
      fireEvent.press(getByTestId('go-back'))
      expect(goBack).toHaveBeenCalled()
    })
  })

  describe('When changing the input of the name field to Pablo', () => {
    it('It should change the value name to Pablo', () => {
      const { getByPlaceholderText } = render(component)
      const newNameValue = 'Pablo'
      const inputName = getByPlaceholderText('Name')
      fireEvent.changeText(inputName, newNameValue)
      expect(inputName.props.value).toBe(newNameValue)
    })
  })

  describe('When changing the input of the email to pablo@gmail.com', () => {
    it('It should change the value email to pablo@gmail.com', () => {
      const { getByPlaceholderText } = render(component)
      const newEmailValue = 'pablo@gmail.com'
      const inputEmail = getByPlaceholderText('Email')
      fireEvent.changeText(inputEmail, newEmailValue)
      expect(inputEmail.props.value).toBe(newEmailValue)
    })
  })

  describe('When changing the input of the password to a1234567', () => {
    it('It should change the value password to a1234567', () => {
      const { getByPlaceholderText } = render(component)
      const newPasswordValue = 'a1234567'
      const inputPassword = getByPlaceholderText('Password')
      fireEvent.changeText(inputPassword, newPasswordValue)
      expect(inputPassword.props.value).toBe(newPasswordValue)
    })
  })

  describe('When changing the input of the repeate password to a1234567', () => {
    it('It should change the value repeated password to a1234567', () => {
      const { getByPlaceholderText } = render(component)
      const newRepeatPasswordValue = 'a1234567'
      const inputRepeatPassword = getByPlaceholderText('Repeat password')
      fireEvent.changeText(inputRepeatPassword, newRepeatPasswordValue)
      expect(inputRepeatPassword.props.value).toBe(newRepeatPasswordValue)
    })
  })

  describe('When pressing on SUBMIT with an invalid email', () => {
    it('It should call the alert with \'Please add a valid email\' ', () => {
      const { getByTestId, getByPlaceholderText } = render(component)
      global.alert = jest.fn()
      const newNameValue = 'Pablo'
      const inputName = getByPlaceholderText('Name')
      fireEvent.changeText(inputName, newNameValue)
      const newEmailValue = 'pablomail.com'
      const inputEmail = getByPlaceholderText('Email')
      fireEvent.changeText(inputEmail, newEmailValue)
      const newPasswordValue = 'a1234567'
      const inputPassword = getByPlaceholderText('Password')
      fireEvent.changeText(inputPassword, newPasswordValue)
      const newRepeatPasswordValue = 'a1234567'
      const inputRepeatPassword = getByPlaceholderText('Repeat password')
      fireEvent.changeText(inputRepeatPassword, newRepeatPasswordValue)
      fireEvent.press(getByTestId('valid-input'))
      expect(global.alert).toHaveBeenCalledWith('Please add a valid email')
    })
  })

  describe('When pressing on SUBMIT with an invalid password', () => {
    it('It should call the alert with \'Please add a valid password! Minimum of 8 characters and at least one letter\' ', () => {
      const { getByTestId, getByPlaceholderText } = render(component)
      global.alert = jest.fn()
      const newNameValue = 'Pablo'
      const inputName = getByPlaceholderText('Name')
      fireEvent.changeText(inputName, newNameValue)
      const newEmailValue = 'pablo@gmail.com'
      const inputEmail = getByPlaceholderText('Email')
      fireEvent.changeText(inputEmail, newEmailValue)
      const newPasswordValue = '34567'
      const inputPassword = getByPlaceholderText('Password')
      fireEvent.changeText(inputPassword, newPasswordValue)
      const newRepeatPasswordValue = 'a1234567'
      const inputRepeatPassword = getByPlaceholderText('Repeat password')
      fireEvent.changeText(inputRepeatPassword, newRepeatPasswordValue)
      fireEvent.press(getByTestId('valid-input'))
      expect(global.alert).toHaveBeenCalledWith('Please add a valid password! Minimum of 8 characters and at least one letter')
    })
  })

  describe('When pressing on SUBMIT with valid inputs', () => {
    it('It should call the alert with \'Thank you for joining Fairhand!\' ', () => {
      const { getByTestId, getByPlaceholderText } = render(component)
      global.alert = jest.fn()
      const newNameValue = 'Pablo'
      const inputName = getByPlaceholderText('Name')
      fireEvent.changeText(inputName, newNameValue)
      const newEmailValue = 'pablo@gmail.com'
      const inputEmail = getByPlaceholderText('Email')
      fireEvent.changeText(inputEmail, newEmailValue)
      const newPasswordValue = 'a1234567'
      const inputPassword = getByPlaceholderText('Password')
      fireEvent.changeText(inputPassword, newPasswordValue)
      const newRepeatPasswordValue = 'a1234567'
      const inputRepeatPassword = getByPlaceholderText('Repeat password')
      fireEvent.changeText(inputRepeatPassword, newRepeatPasswordValue)
      fireEvent.press(getByTestId('valid-input'))
      expect(global.alert).toHaveBeenCalledWith('Thank you for joining Fairhand!')
    })

    it('It should call navigation.navigate', () => {
      component = (
        <Provider store={mockStore({
          userReducer: {
            user: {
              email: 'pablo@gmail.com'
            }
          }
        })}><Register navigation={{ goBack, navigate }}/></Provider>)
      render(component)
      expect(navigate).toHaveBeenCalledWith('TabNavigator')
    })
  })
})
