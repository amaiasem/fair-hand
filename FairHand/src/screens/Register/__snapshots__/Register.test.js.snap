// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a component Login It should match the snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Memo(Connect(Register))
    navigation={
      Object {
        "goBack": [MockFunction],
        "navigate": [MockFunction],
      }
    }
  />
</Provider>
`;
